services/repository/src/test/java/org/sagebionetworks/repo/web/service/EntityServiceImplUnitTest.java

```diff
@@ -235,4 +248,85 @@ public void testGetActionRequiredForDownload() {
 		
 		verify(mockAuthManager).getActionsRequiredForDownload(userInfo, List.of(123L));
 	}
+
+	@Test
+	public void testValidateDefiningSql() {
+		String sql = "select * from syn123";
+		DefiningSqlEntityType definingSqlEntityType = DefiningSqlEntityType.materializedview;
+		EntityType entityType = EntityType.materializedview;
+		ValidateDefiningSqlRequest mockRequest = new ValidateDefiningSqlRequest().setDefiningSql(sql).setEntityType(definingSqlEntityType);
+
+		when(mockMetadataProviderFactory.getMetadataProvider(any())).thenReturn(materializedViewProviders);
+
+		// Call under test
+		ValidateDefiningSqlResponse response = entityService.validateDefiningSql(mockRequest);
+
+		verify(mockMetadataProviderFactory).getMetadataProvider(entityType);
+		verify(mockMaterializedViewDefiningSqlProvider).validateDefiningSql(sql);
+
+		assertTrue(response.getIsValid());
+		assertNull(response.getInvalidReason());
+	}
+
+	@Test
+	public void testValidateDefiningSqlWithNullProviders() {
+		String sql = "select * from syn123";
+		DefiningSqlEntityType definingSqlEntityType = DefiningSqlEntityType.materializedview;
+		EntityType entityType = EntityType.materializedview;
+		ValidateDefiningSqlRequest mockRequest = new ValidateDefiningSqlRequest().setDefiningSql(sql).setEntityType(definingSqlEntityType);
+
+		when(mockMetadataProviderFactory.getMetadataProvider(any())).thenReturn(null);
+
+		// Call under test
+		ValidateDefiningSqlResponse response = entityService.validateDefiningSql(mockRequest);
+
+		verify(mockMetadataProviderFactory).getMetadataProvider(entityType);
+		verify(mockMaterializedViewDefiningSqlProvider, never()).validateDefiningSql(sql);
+
+		assertFalse(response.getIsValid());
+		assertEquals("No provider found for entity type " + entityType, response.getInvalidReason());
+	}
+
+	@Test 
+	public void testValidateDefiningSqlWithInvalidProviders() {
+		String sql = "select * from syn123";
+		DefiningSqlEntityType definingSqlEntityType = DefiningSqlEntityType.materializedview;
+		EntityType entityType = EntityType.materializedview;
+		ValidateDefiningSqlRequest mockRequest = new ValidateDefiningSqlRequest().setDefiningSql(sql).setEntityType(definingSqlEntityType);
+		ArrayList<EntityProvider<? extends Entity>> invalidProviders = new ArrayList<>();
+		invalidProviders.add(mockProjectCreateProvider);
+
+		when(mockMetadataProviderFactory.getMetadataProvider(any())).thenReturn(invalidProviders);
+
+		// Call under test
+		ValidateDefiningSqlResponse response = entityService.validateDefiningSql(mockRequest);
+
+		verify(mockMetadataProviderFactory).getMetadataProvider(entityType);
+		verify(mockMaterializedViewDefiningSqlProvider, never()).validateDefiningSql(sql);
+
+		assertFalse(response.getIsValid());
+		assertEquals("No provider found for entity type " + entityType, response.getInvalidReason());
```

Comment:
Two cases:
No provider for key.
Have a provider but does not implement interface
Both should throw as 500 (IllegalStateException()